File: PROG2-MIDTERM/MidtermLabWork2/csv_processor.js
/**
 * Midterm Lab Work 2: CSV Dataset Processing
 * Student: Lord Clyde W. Pega
 * Problems: MP01, MP11, MP19
 */

const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question('Enter the dataset file path: ', (filePath) => {
    try {
        const path = filePath.replace(/['"]+/g, '');
        const content = fs.readFileSync(path, 'utf8');
        const allLines = content.split(/\r?\n/);

        // Filter out empty lines and skip the 6 metadata lines
        const dataRows = allLines.filter(line => line.trim() !== '').slice(6);
        
        // Map rows into arrays
        const dataset = dataRows.map(row => row.split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/));

        console.log("\n====================================================");
        console.log("   LAB WORK 2 RESULTS (JS) - LORD CLYDE W. PEGA");
        console.log("====================================================");

        // MP01: Total Records
        const totalRecords = dataset.length - 1; 
        console.log(`[MP01] Total Student Records: ${totalRecords}`);

        // MP11: Frequency Count (Exam Type - Column Index 3)
        const freq = {};
        for (let i = 1; i < dataset.length; i++) {
            const exam = dataset[i][3]?.trim() || "Unknown";
            freq[exam] = (freq[exam] || 0) + 1;
        }
        console.log("\n[MP11] Frequency Count (Exam Type):");
        Object.entries(freq).forEach(([key, val]) => {
            console.log(`   - ${key.padEnd(15)} : ${val}`);
        });

        // MP19: Summary Report
        console.log("\n[MP19] Dataset Summary Report:");
        console.log(`   - Source: ${path}`);
        console.log(`   - Total File Lines: ${allLines.length}`);
        console.log(`   - Processed Records: ${totalRecords}`);
        console.log("====================================================\n");

    } catch (err) {
        console.log("Error: Could not process file.");
    } finally {
        rl.close();
    }
});

File: PROG2-MIDTERM/MidtermLabWork2/CSVProcessor.java
/**
 * Midterm Lab Work 2: CSV Dataset Processing
 * Student: Lord Clyde W. Pega
 * Problems: MP01, MP11, MP19
 */
import java.io.*;
import java.util.*;

public class CSVProcessor {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the dataset file path: ");
        String filePath = scanner.nextLine().replace("\"", "");

        List<String[]> records = new ArrayList<>();
        int lineCount = 0;

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            // Skip 6 lines of metadata
            for (int i = 0; i < 6; i++) { 
                br.readLine(); 
                lineCount++; 
            }

            while ((line = br.readLine()) != null) {
                lineCount++;
                if (line.trim().isEmpty()) continue;
                
                // Regex to split by comma but ignore commas inside quotes
                String[] values = line.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
                records.add(values);
            }

            // Display Results
            printResults(records, lineCount, filePath);

        } catch (Exception e) {
            System.out.println("Error: File not found or could not be read.");
        } finally {
            scanner.close();
        }
    }

    public static void printResults(List<String[]> data, int totalLines, String path) {
        String dash = "=".repeat(52);
        System.out.println("\n" + dash);
        System.out.println("   LAB WORK 2 RESULTS - LORD CLYDE W. PEGA");
        System.out.println(dash);

        // MP01: Total Records (minus header)
        int totalRecords = data.size() - 1;
        System.out.println("[MP01] Total Student Records: " + totalRecords);

        // MP11: Frequency Count for Exam Type (Column Index 3)
        Map<String, Integer> freqMap = new HashMap<>();
        for (int i = 1; i < data.size(); i++) {
            String exam = data.get(i)[3].trim();
            freqMap.put(exam, freqMap.getOrDefault(exam, 0) + 1);
        }
        System.out.println("\n[MP11] Frequency Count (Exam Type):");
        for (Map.Entry<String, Integer> entry : freqMap.entrySet()) {
            System.out.printf("   - %-15s : %d%n", entry.getKey(), entry.getValue());
        }

        // MP19: Summary Report
        System.out.println("\n[MP19] Dataset Summary Report:");
        System.out.println("   - Source Path: " + path);
        System.out.println("   - File Line Count: " + totalLines);
        System.out.println("   - Valid Records: " + totalRecords);
        System.out.println(dash + "\n");
    }
}